plugin/src/main/java/org/opensearch/ml/cluster/DiscoveryNodeHelper.java

```diff
@@ -68,6 +78,26 @@ public DiscoveryNode[] getEligibleNodes() {
         }
     }
 
+    public String[] filterEligibleNodes(String[] nodeIds) {
+        if (nodeIds == null || nodeIds.length == 0) {
+            return nodeIds;
+        }
+        DiscoveryNode[] nodes = getNodes(nodeIds);
+        final List<String> eligibleNodes = new ArrayList<>();
+        for (DiscoveryNode node : nodes) {
+            if (excludedNodeNames != null && excludedNodeNames.contains(node.getName())) {
+                continue;
+            }
+            if (MLNodeUtils.isMLNode(node)) {
+                eligibleNodes.add(node.getId());
+            }
+            if (!onlyRunOnMLNode && node.isDataNode() && isEligibleDataNode(node)) {
+                eligibleNodes.add(node.getId());
+            }
```

Comment:
If a node has both ml and data roles, eligibleNodes could have duplicated nodes. Should we avoid the duplicate?

Reply:
Good catch, will fix